### PR TITLE
Improve FMHA bwd

### PR DESCRIPTION
**Pull request by rocking5566**
This PR integrate the bwd optimization
https://github.com/ROCm/composable_kernel/pull/1397

### CODE REVIEW

File: csrc/flash_attn_ck/mha_varlen_bwd.cpp [hunk 1]
```diff
@@ -148,6 +173,10 @@ fmha_bwd_args get_ck_fmha_varlen_bwd_args(const mask_info &mask,
                          0, // nhead_stride_randval
                          nhead_stride_do,
                          nhead_stride_lse,
+                         nhead_stride_dq_acc,
+                         nhead_stride_q,
```
Comment (danyao12) [inline csrc/flash_attn_ck/mha_varlen_bwd.cpp:177]: nhead_stride_dq/nhead_stride_dk/nhead_stride_dv

File: csrc/flash_attn_ck/mha_varlen_bwd.cpp [hunk 2]
```diff
@@ -308,13 +339,20 @@ mha_varlen_bwd(const at::Tensor &dout,                   // total_q x num_heads
...
 
     auto opts = q.options();
     auto softmax_d = torch::empty({batch_size, num_heads, max_seqlen_q}, opts.dtype(at::kFloat));
-    // TODO - CK does not support dq_accum
+    at::Tensor dq_accum;
+
+    if (!deterministic) {
+        dq_accum = torch::zeros({1, total_q, num_heads, head_size_8x}, opts.dtype(at::kFloat));
+    } else {
+        const ck_tile::index_t kN0 = (head_size_8x > 32 & head_size_8x <= 128) ? 128 : 64;
```
Comment (danyao12) [inline csrc/flash_attn_ck/mha_varlen_bwd.cpp:352]: const ck_tile::index_t kN0 = (head_size_8x <= 128) ? 128 : 64;

File: tests/test_flash_attn_ck.py [hunk 3]
```diff
@@ -24,7 +24,7 @@
 )
 
 def is_bwd_hdim_supported(d):
-    return d <= 128 and d % 2 == 0
+    return d <= 256
 
 
 def ck_randval_to_dropout_mask(randval, p):
```
Comment (danyao12) [inline tests/test_flash_attn_ck.py:30]: return torch.floor(255.0 * (1 - p)) - randval

File: csrc/flash_attn_ck/mha_bwd.cpp [hunk 4]
```diff
@@ -143,6 +167,10 @@ fmha_bwd_args get_ck_fmha_bwd_args(const mask_info &mask,
                          0, // nhead_stride_randval
                          nhead_stride_do,
                          nhead_stride_lse,
+                         nhead_stride_dq_acc,
+                         nhead_stride_q,
```
[the remark this answers is not in the document]
Comment (rocking5566) [reply csrc/flash_attn_ck/mha_bwd.cpp]: good catch!
5346b5b9f6b9980b61ef11f389d82f758a009eb8

File: csrc/flash_attn_ck/mha_varlen_bwd.cpp (hunk 1 above)
Comment (rocking5566) [reply csrc/flash_attn_ck/mha_varlen_bwd.cpp]: 5346b5b9f6b9980b61ef11f389d82f758a009eb8

File: csrc/flash_attn_ck/mha_varlen_bwd.cpp (hunk 2 above)
Comment (rocking5566) [reply csrc/flash_attn_ck/mha_varlen_bwd.cpp]: 5346b5b9f6b9980b61ef11f389d82f758a009eb8

File: csrc/flash_attn_ck/mha_varlen_bwd.cpp [hunk 5]
```diff
@@ -50,45 +54,62 @@ fmha_bwd_args get_ck_fmha_varlen_bwd_args(const mask_info &mask,
...
     ck_tile::index_t batch_stride_v = 0;
+    ck_tile::index_t stride_v = v.stride(0);
+    ck_tile::index_t nhead_stride_v = v.stride(1);
+
+    // o: (total_q, nheads, hdim)
     ck_tile::index_t batch_stride_o = 0;
+    ck_tile::index_t stride_o = out.stride(0);
+    ck_tile::index_t nhead_stride_o = out.stride(1);
+
+    // lse: (batch_size, nheads, max_seqlen_q)
```
Comment (danyao12) [inline csrc/flash_attn_ck/mha_varlen_bwd.cpp:80]: // lse: (nheads, total_q)

Comment (rocking5566) [reply csrc/flash_attn_ck/mha_varlen_bwd.cpp]: 2c80b86591f6c8d32a054fd9fa677ac6f8c483c6

File: tests/test_flash_attn_ck.py (hunk 3 above)
Comment (rocking5566) [reply tests/test_flash_attn_ck.py:31]: d2ed41392e67eeb25ec6aa0c8e7bd88cc36741e4

**Pull request merged**